presentation-api/controlling-ua/reconnectToPresentation_success-manual.html

```diff
@@ -1,72 +1,49 @@
-        return request.start()
-                .then(function (connection) {
-                    // save the presentation id within the
-                    // local storage of the browser
-                    localStorage["presId"] = connection.id;
+    async_test(function (t) {
+        var startBtn = document.getElementById("startBtn");
+        startBtn.onclick = t.step_func(function () {
+            startBtn.disabled = true;
+            var request = new PresentationRequest(presentationUrls);
+            // start presentation
+            request.start().then(t.step_func(function (oldConnection) {
+                // presentationId will be used to reconnect to presentation
+                var presentationId = oldConnection.id;
+                var oldConnectionClosed = false;
+                oldConnection.onclose = t.step_func(function () {
+                    oldConnectionClosed = true;
+                    // reconnect to presentation using presentationId
+                    request.reconnect(presentationId).then(t.step_func(function (newConnection) {
+                        assert_equals(newConnection.state, 'connecting',"connection should be in 'connecting' state");
+                        assert_equals(newConnection.id, presentationId, "Ids of old and new connections must be equal");
+                        t.done();
+                    })).catch(t.step_func(function () {
+                        assert_unreached("Reconnect to presentation promise rejected");
+                    }));
                 });
-    }
-
-    // ------------------------------------------------
-    // Reconnect to Presentation Test (success) - begin
-    // ------------------------------------------------
-    var reconnectToPresentation = function () {
-
-        promise_test(function () {
-
-            // get the saved presentation id
-            // for the reconnect function if exist
-            if (localStorage["presId"]) var presId = localStorage["presId"];
-
-            return request.reconnect(presId)
-                    .then(function (connection) {
-                        assert_true(connection instanceof PresentationConnection);
-                        done();
-                    });
-        }, 'The reconnection was successful.');
-    }
-    // ----------------------------------------------
-    // Reconnect to Presentation Test (success) - end
-    // ----------------------------------------------
+                // close connection. test continue after onclose event is fired.
+                oldConnection.close();
+                // the test fails if after 5s the close event is not fired
+                t.step_timeout(function () {
+                    !oldConnectionClosed && assert_unreached("Close event on old connection not fired");
+                },5000);
+            })).catch(t.step_func(function () {
+                assert_unreached("Error on start presentation");
+            }));
```

Comment:
Thinking about this again, the test harness `EventWatcher` class could further help streamline the above code (avoiding the call to `new Promise` in particular):

```js
function startPresentation() {
    document.getElementById("startBtn").disabled = true;

    promise_test(function (t) {
        var request = new PresentationRequest(presentationUrls);
        var presentationId = null;
        return request.start()
            .then(function (connection) {
                presentationId = connection.id;

                // No more user input needed, re-enable test timeout
                t.step_timeout(function () {
                    t.force_timeout();
                    t.done();
                }, 5000);

                // Close connection and wait for "close" event
                connection.close();
                var eventWatcher = new EventWatcher(t, connection, 'close');
                return eventWatcher.wait_for('close');
            })
            .then(function () {
                // Connection now closed, let's reconnect to it
                return request.reconnect(presentationId);
            })
            .then(function (connection) {
                assert_equals(connection.state, "connecting", "connection should be in 'connecting' state");
                assert_equals(connection.id, presentationId, "Ids of old and new connections must be equal");
            });
    });
}
```